runtime/primitives/src/xcm.rs

```diff
@@ -0,0 +1,233 @@
+// Copyright 2020-2021 Manta Network.
+// This file is part of Manta.
+//
+// Manta is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Manta is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Manta.  If not, see <http://www.gnu.org/licenses/>.
+
+use sp_std::marker::PhantomData;
+use sp_runtime::traits::{Convert, Zero};
+
+use frame_support::{weights::{Weight, constants::WEIGHT_PER_SECOND}};
+use frame_support::traits::fungibles::Mutate;
+use frame_support::pallet_prelude::Get;
+
+use xcm::v1::{
+	AssetId as xcmAssetId,
+	Junction::{AccountId32, Parachain},
+	Junctions::*,
+	MultiAsset, MultiLocation, NetworkId, Fungibility
+};
+use xcm::latest::Error as XcmError;
+use xcm_builder::TakeRevenue;
+use xcm_executor::traits::{FilterAssetLocation, WeightTrader, MatchesFungibles};
+use crate::{UnitsToWeightRatio, AssetIdLocationGetter};
+
+
+pub trait Reserve {
+	/// Returns assets reserve location.
+	fn reserve(&self) -> Option<MultiLocation>;
+}
+
+// Takes the chain part of a MultiAsset
+impl Reserve for MultiAsset {
+	fn reserve(&self) -> Option<MultiLocation> {
+		// We only care about concrete location now.
+		if let xcmAssetId::Concrete(location) = self.id.clone() {
+			let first_interior = location.first_interior();
+			let parents = location.parent_count();
+			match (parents, first_interior.clone()) {
+				(0, Some(Parachain(id))) => Some(MultiLocation::new(0, X1(Parachain(id.clone())))),
+				(1, Some(Parachain(id))) => Some(MultiLocation::new(1, X1(Parachain(id.clone())))),
+				(1, _) => Some(MultiLocation::parent()),
+				_ => None,
+			}
+		} else {
+			None
+		}
```

Comment:
Are all these clones needed? I suspect at least the clones on lines 49 and 50 are not needed.